src/interpreter.rs

```diff
@@ -1256,69 +1256,99 @@ impl<'source> Interpreter<'source> {
-        };
+            if args.len() != params.len() {
+                return Err(span.source.error(
+                    span.line,
+                    span.col,
+                    format!(
+                        "mismatch in number of arguments. supplied {}, expected {}",
+                        params.len(),
+                        args.len()
+                    )
+                    .as_str(),
+                ));
+            }
 
-        // Back up local variables of current function and empty
-        // the local variables of callee function.
-        let scopes = std::mem::take(&mut self.scopes);
+            let mut args_scope = Scope::new();
+            for (idx, a) in args.iter().enumerate() {
+                let a = match a {
+                    Expr::Var(s) => s.text(),
+                    _ => unimplemented!("destructuring function arguments"),
+                };
+                //TODO: check call in params
+                args_scope.insert(a.to_string(), self.eval_expr(&params[idx])?);
+            }
+
+            let ctx = Context {
+                key_expr: None,
+                output_expr,
+                value: Value::new_set(),
+            };
 
-        // Set the arguments scope.
-        self.scopes.push(args_scope);
-        let value = self.eval_rule_bodies(ctx, span, bodies)?;
-        let result = match &value {
-            Value::Set(s) if s.len() == 1 => Ok(s.iter().next().unwrap().clone()),
-            Value::Set(s) if !s.is_empty() => Err(span.source.error(
+            // Back up local variables of current function and empty
+            // the local variables of callee function.
+            let scopes = std::mem::take(&mut self.scopes);
+
+            // Set the arguments scope.
+            self.scopes.push(args_scope);
+            let value = self.eval_rule_bodies(ctx, span, bodies)?;
+            let result = match &value {
+                Value::Set(s) if s.len() == 1 => s.iter().next().unwrap().clone(),
+                Value::Set(s) if !s.is_empty() => {
+                    return Err(span.source.error(
+                        span.line,
+                        span.col,
+                        format!("function produced multiple outputs {value:?}").as_str(),
+                    ))
+                }
+                // If the function successfully executed, but did not return any value, then return true.
+                Value::Set(s) if s.is_empty() && output_expr.is_none() => Value::Bool(true),
+
+                // If the function execution resulted in undefined, then propagate it.
+                Value::Undefined => Value::Undefined,
+                _ => bail!("internal error: function did not return set {value:?}"),
+            };
+
+            // Restore local variables for current context.
+            self.scopes = scopes;
+
+            // If the function successfully executed and returned true, not necessary to execute the other
```

Comment:
This assumption may not be correct. It seems that Rego requires all the definitions to be executed even if one definition returned true.

Reply:
You're absolutely right, I think I mixed up two different rego objects: "incremental rules" and "multiple function definitions" (I told you I was new to rego !). 

To my understanding, rules with the same name should be executed one after the another until true is returned, but for functions, as you said, all functions should be executed, and the return value should be the same, regardless of the value set (defined as in your example).

I will fix that tomorrow